import/index_java/src/org/vufind/index/UpdateDateTracker.java

```diff
@@ -69,37 +65,49 @@ private void createRow(Timestamp newRecordChange) throws SQLException
         lastRecordChange = newRecordChange;
 
         // Save new values to the database:
-        insertSql.setString(1, core);
-        insertSql.setString(2, id);
-        insertSql.setTimestamp(3, firstIndexed);
-        insertSql.setTimestamp(4, lastIndexed);
-        insertSql.setTimestamp(5, lastRecordChange);
-        insertSql.executeUpdate();
+        try (
+            PreparedStatement insertSql = db.prepareStatement(
+                "INSERT INTO change_tracker(core, id, first_indexed, last_indexed, last_record_change) " +
+                "VALUES(?, ?, ?, ?, ?);"
+            );
+        ) {
+            insertSql.setString(1, core);
+            insertSql.setString(2, id);
+            insertSql.setTimestamp(3, firstIndexed);
+            insertSql.setTimestamp(4, lastIndexed);
+            insertSql.setTimestamp(5, lastRecordChange);
+            insertSql.executeUpdate();
+        }
     }
 
     /* Private support method: read a row from the change_tracker table.
      */
     private boolean readRow() throws SQLException
     {
-        selectSql.setString(1, core);
-        selectSql.setString(2, id);
-        ResultSet result = selectSql.executeQuery();
-
-        // No results?  Free resources and return false:
-        if (!result.first()) {
-            result.close();
-            return false;
+        boolean returnValue = true;
+        try (
+            PreparedStatement selectSql = db.prepareStatement(
+                "SELECT first_indexed, last_indexed, last_record_change, deleted " +
+                "FROM change_tracker WHERE core = ? AND id = ?;",
+                ResultSet.TYPE_SCROLL_INSENSITIVE, ResultSet.CONCUR_READ_ONLY
+            )
+        ) {
+            selectSql.setString(1, core);
+            selectSql.setString(2, id);
+            try (ResultSet result = selectSql.executeQuery()) {
+                // No results? Return false:
+                if (!result.first()) {
+                    returnValue = false;
```

Comment:
It looks like you should still be safe just calling `return false` here; the close would still work.  https://stackoverflow.com/questions/22947755/try-with-resources-and-return-statements-in-java

Reply:
Thanks! I've simplified as you suggested and it still seems to be working as expected for me.